modules/distribution-zones/src/test/java/org/apache/ignite/internal/distributionzones/DistributionZoneManagerFilterTest.java

```diff
@@ -0,0 +1,212 @@
+    );
+
+    private static final LogicalNode D = new LogicalNode(
+            new ClusterNode("4", "D", new NetworkAddress("localhost", 123)),
+            Map.of("region", "CN", "storage", "SSD", "dataRegionSize", "20")
+    );
+
+    @Test
+    void testFilterOnStart() throws Exception {
+        String filter = "$[?(@.storage == 'SSD' || @.region == 'US')]";
+
+        topology.putNode(A);
+        topology.putNode(B);
+        topology.putNode(C);
+
+        Set<LogicalNode> clusterNodes = Set.of(A, B, C);
+
+        mockVaultZonesLogicalTopologyKey(clusterNodes, vaultMgr);
+
+        startDistributionZoneManager();
+
+        distributionZoneManager.createZone(
+                new DistributionZoneConfigurationParameters.Builder(ZONE_NAME)
+                        .dataNodesAutoAdjustScaleUp(IMMEDIATE_TIMER_VALUE)
+                        .dataNodesAutoAdjustScaleDown(IMMEDIATE_TIMER_VALUE)
+                        .filter(filter)
+                        .build()
+        ).get(10_000, TimeUnit.MILLISECONDS);
+
+        Set<String> nodes = distributionZoneManager.topologyVersionedDataNodes(
+                1,
+                topology.getLogicalTopology().version()
+        ).get(10_000, TimeUnit.MILLISECONDS);
+
+        assertEquals(Set.of(A, C).stream().map(ClusterNode::name).collect(Collectors.toSet()), nodes);
+    }
+
+    @Test
+    void testFilterOnScaleUp() throws Exception {
+        String filter = "$[?(@.storage == 'SSD' || @.region == 'US')]";
+
+        topology.putNode(A);
+        topology.putNode(B);
+        topology.putNode(C);
+
+        Set<LogicalNode> clusterNodes = Set.of(A, B, C);
+
+        mockVaultZonesLogicalTopologyKey(clusterNodes, vaultMgr);
+
+        startDistributionZoneManager();
+
+        distributionZoneManager.createZone(
+                new DistributionZoneConfigurationParameters.Builder(ZONE_NAME)
+                        .dataNodesAutoAdjustScaleUp(IMMEDIATE_TIMER_VALUE)
+                        .dataNodesAutoAdjustScaleDown(IMMEDIATE_TIMER_VALUE)
+                        .filter(filter)
+                        .build()
+        ).get(10_000, TimeUnit.MILLISECONDS);
+
+        Set<String> nodes = distributionZoneManager.topologyVersionedDataNodes(
+                1,
+                topology.getLogicalTopology().version()
+        ).get(10_000, TimeUnit.MILLISECONDS);
+
+        assertEquals(Set.of(A, C).stream().map(ClusterNode::name).collect(Collectors.toSet()), nodes);
+
+        topology.putNode(D);
+
+        nodes = distributionZoneManager.topologyVersionedDataNodes(
+                1,
+                topology.getLogicalTopology().version()
+        ).get(10_000, TimeUnit.MILLISECONDS);
+
+        assertEquals(Set.of(A, C, D).stream().map(ClusterNode::name).collect(Collectors.toSet()), nodes);
+    }
+
+    @Test
+    void testFilterOnScaleDown() throws Exception {
+        String filter = "$[?(@.storage == 'SSD' || @.region == 'US')]";
```

Comment:
Seems that we have
```
        String filter = "$[?(@.storage == 'SSD' || @.region == 'US')]";

        topology.putNode(A);
        topology.putNode(B);
        topology.putNode(C);

        Set<LogicalNode> clusterNodes = Set.of(A, B, C);

        mockVaultZonesLogicalTopologyKey(clusterNodes, vaultMgr);

        startDistributionZoneManager();

        distributionZoneManager.createZone(
                new DistributionZoneConfigurationParameters.Builder(ZONE_NAME)
                        .dataNodesAutoAdjustScaleUp(IMMEDIATE_TIMER_VALUE)
                        .dataNodesAutoAdjustScaleDown(IMMEDIATE_TIMER_VALUE)
                        .filter(filter)
                        .build()
        ).get(10_000, TimeUnit.MILLISECONDS);

        Set<String> nodes = distributionZoneManager.topologyVersionedDataNodes(
                1,
                topology.getLogicalTopology().version()
        ).get(10_000, TimeUnit.MILLISECONDS);

        assertEquals(Set.of(A, C).stream().map(ClusterNode::name).collect(Collectors.toSet()), nodes);
```
In many tests, e.g. testFilterOnStart, testFilterOnScaleUp, testFilterOnScaleUpWithNewAttributesAfterRestart. Should we extract it to some private helper?

Reply:
extracted